### PR TITLE
adding aarch64 build for DockerHub

### PR DESCRIPTION
**Pull request by KurtStam**
#287, Adding support for building an aarch64 based image on Dockerhub. The base aarch64 base image contains a qemu library which is used after issuing the 

RUN [ "cross-build-start" ]

and is disabled after issuing

RUN [ "cross-build-end" ]

This allows building the image for aarch64:
https://hub.docker.com/r/project31/aarch64-docker-openvpn/builds/

Maybe I should merge the two README's by adding a section on for aarch64?

### CODE REVIEW

**Comment by kylemanna** [comment]
I was reviewing the Travis-CI output and it looks like it's running the tests against an x86_64 server and not a QEMU image.  Did you do anything to run the tests via QEMU or similar?

It's essential to get tests integrated to add aarch64 support added.

**Comment by KurtStam** [comment]
Hi @kylemanna, it should be able to build (and run the tests) on a x86_64 machine, as the image contains the QEMU library to translate it to aarch64 (which is what the image is). I've tested this on Dockerhub but not yet on Travis-CI. Hopefully I will have some time soon to verify this!

**Comment by KurtStam** [comment]
@kylemanna I added a second build to travis which adds the aarch64 based image. Although I'm able to run the tests on my aarch64 boards, I can't get the tests to run on travis. In order for it to work on travis as well we'd need them to provide a newer (version 4) kernel. So added some logic to not run the tests for the aarch64 architecture. How about adding a docker push to travis so I could run the tests elsewhere?

**Comment by kylemanna** [comment]
@KurtStam you can create a Travis-CI build on your user account and for development and testing.  That should run identically to my build.  Once you get the kinks worked out, lets review the Travis CI details and we can merge it.

**Comment by KurtStam** [comment]
@kylemanna - got the kinks worked out.

**Comment by kylemanna** [comment]
Cool!

File: .travis.yml [hunk 1]
```diff
@@ -19,10 +23,12 @@ install:
 # Assist with ci test debugging:
 #   - DEBUG=1
 before_script:
-    - image="kylemanna/openvpn"
-    - docker build -t "$image" .
+# prepare qemu
+    - if [ $ARCH == 'aarch64' ]; then docker run --rm --privileged multiarch/qemu-user-static:register --reset; fi
```
Comment (kylemanna) [inline .travis.yml]: What does this do?  Please provide  a comment describing what's happening here.

File: README-aarch64.md [hunk 2]
```diff
@@ -0,0 +1,208 @@
+# OpenVPN for Docker
```
Comment (kylemanna) [inline README-aarch64.md]: Instead of adding a new README, we should put an `aarch64.md` in the `docs/` folder.

File: Dockerfile.aarch64 [hunk 3]
```diff
@@ -1,9 +1,11 @@
 # Original credit: https://github.com/jpetazzo/dockvpn
 
 # Smallest base image
-FROM aarch64/alpine:3.5
+FROM docker.io/project31/aarch64-alpine-qemu:3.5-7
 
-MAINTAINER Kyle Manna <kyle@kylemanna.com>
+RUN [ "cross-build-start" ]
```
Comment (kylemanna) [inline Dockerfile.aarch64]: Why is the cross build necessary?  We should be pulling working binaries from upstream Alpine.

File: .travis.yml [hunk 4]
```diff
@@ -19,10 +23,12 @@ install:
...
+    - image="$IMAGE"
+    - docker build -f $DOCKERFILE -t "$image" .
     - docker inspect "$image"
-    - docker run --rm "$image" openvpn --version || true # why does it return 1?
+    - docker run --rm "$image" openvpn --version || true; # why does it return 1?
```
Comment (kylemanna) [inline .travis.yml]: This line is extraneous

File: Dockerfile.aarch64 [hunk 5]
```diff
@@ -1,9 +1,11 @@
...
+
+MAINTAINER Kurt Stam <kurt.stam@redhat.com>
```
Comment (kylemanna) [inline Dockerfile.aarch64]: This is fine, but can you use the new `LABEL` approach as `MAINTAINER` is apparently deprecated.  It's also possible to specify two names, you should specify both mine and yours assuming you accept the responsibility and agree with my general approach. :)

See https://docs.docker.com/engine/reference/builder/#maintainer-deprecated

File: .travis.yml (hunk 4 above)
Comment (KurtStam) [reply .travis.yml]: it was in the original travis file. I'm happy to remove it.

**Comment by KurtStam** [comment]
@kylemanna, I made things quite a bit nicer.

1. It's now running of the same Dockerfile since recent, the FROM can be parameterized. So I'm taking advantage of that.

2. FROM images that are NOT x86_64 need to have the qemu inside of the base image. The multiarch/alpine images have them for a variety of architectures.

3. Travis now builds images for amd64 and arm64 architectures and pushed them to docker hub with tags amd64-latest and arm-latest: see https://hub.docker.com/r/project31/docker-openvpn/tags/

4. I introduced a new step on travis that, after the docker builds are ready, creates a manifest. The manifest is a new docker feature that has references to both amd64 and arm64 images. This way the README.txt can stay untouched. The user can still execute the commands listed and docker will figure out which layers to download for the specific architecture. The travis build is pushing to your dockerhub which is failing at the moment (It's successful here for the project31 organization: https://travis-ci.org/PiBeach/docker-openvpn/builds/285868889)

5. The multiarch alpine images have GMT set which makes the official-images test fail as it expects UTC. This is why I had to add `apk add -U tzdata && setup-timezone -z UTC` to the Dockerfile.

I think it's pretty slick now! 

Cheers,

p.s. adding other architectures should now be as easy as pie.

**Comment by miesgre** [comment]
Any news on this? it would be wonderful to have this container to run on other archs

**Comment by KurtStam** [comment]
@miesgre It's good to go.. The images would be uploaded to Kyle's dockerhub when merged but for now you can try https://hub.docker.com/r/project31/aarch64-docker-openvpn/tags/ 

@kylemanna let me know if there is anything else I can do to help.

**Comment by miesgre** [comment]
@KurtStam I tested your images but for armhf (using multiarch/alpine:armhf-latest-stable image as base) and it works! Thanks!!

@kylemanna Please merge this PR (this will solve also #319)

**Comment by kylemanna** [comment]
This can be be merged after the docker login and docker push commands in .tavis.yml are disabled or removed.  They currently break the build and aren't needed for DockerHub integration.

Refer to https://travis-ci.org/kylemanna/docker-openvpn/builds/285870914

Thanks for helping test!

**Comment by KurtStam** [comment]
@kylemanna Changes as requested but I don't really understand how this can work w/o the login and push. Build seems to be failing because of this now.

**Comment by kylemanna** [comment]
All the builds are built by DockerHub as "Automated Builds".  Guides to this are at: https://docs.docker.com/docker-hub/builds/

Instead of building and pushing by CI, Docker is notified of changes to the repo by GitHub and then triggers a build on DockerHub of the latest repo.

**Comment by KurtStam** [comment]
@kylemanna. 

1. OK. Looking at the builds it looks like we are getting bitting by bad signatures in the alpine images. I will look into that some more.

2. To build a manifest image, the images it references to need to have been build. I'm not sure how docker hub supports a this workflow where you build the architecture specific images and then the manifest. Would you consider having travis build the images instead?

**Comment by KurtStam** [comment]
ad 1. Setting the alpine version to v3.6 builds both images. However when building the manifest image  travis reaches out to the actual images on dockerhub it thinks have been build. So that still fails.

ad 2. the build would work if we simply switch to using builds build by travis rather then docker hub.

**Comment by miesgre** [comment]
The automated build feature on the official Docker Hub only runs the normal x86_64 docker implementation. 

I think the way to go is to use travis to build the images. You can make travis to [deploy](https://docs.travis-ci.com/user/deployment) the images to docker hub only when a tag is created. Something like:
```
deploy:      
  - provider: script
    script: sh ./.travis_scripts/upload_to_docker_hub.sh
    skip_cleanup: true
    on:
      tags: true
```
and in `./.travis_scripts/upload_to_docker_hub.sh` you can access to env vars like `${TRAVIS_BUILD_NUMBER}`, `${TRAVIS_COMMIT}` or `${TRAVIS_TAG}`

And the login username and password can be secured using travis [encryption keys](https://docs.travis-ci.com/user/encryption-keys/).

**Comment by KurtStam** [comment]
I've reverted my PR, and am using alpine v.3.7 to avoid the bad signatures in the latest-stable. @kylemanna as @miesgre states you can set your docker credentials as travis encryption keys. That should be the only thing you need to do to make the builds pass.

**Comment by buchdag** [comment]
I really don't like the idea of having to rely on a third party binary to build the manifest, can't this just wait until the `manifest` command has been added to the docker cli ?

The `docker login` and `docker push` are in the same stage as the rest of the test, doesn't that mean that even a build that fails the test get pushed to dockerhub ?

The whole deployment (docker login, docker push, manifest creation) should be a stage of its own separate from the test and should be conditional with `if: branch = master`.

**Comment by miesgre** [comment]
I agree with @buchdag. We should have tree stages:
- test
  Should build the image and test it (in every platform, using the [build matrix](https://docs.travis-ci.com/user/customizing-the-build/#Build-Matrix)
- deploy 
   Should deploy the images to docker hub (only when a tag is created for example..) using the build matrix to upload all archs
- manifest
   Build the manifest

**Comment by KurtStam** [comment]
@buchdag You will have to wait till dockerhub 
- runs on newer kernels so it supports using qemu to build the alternative architectures, and
- until they support some sort of workflow or build dependency at a minimum.

@miesgre the current travis build is meant to generate the 'latest' snapshots. I agree we should add support for official tagged releases.

**Comment by buchdag** [comment]
@KurtStam I'm not saying we should wait until dockerhub can handle the multi arch automated build all by itself (cause it probably won't happen anytime soon) but that, if the solution of having Travis build and push both the images and the manifest is accepted by @kylemanna , I'd prefer that it does not rely on a third party **binary** (project31/docker-manifest) of which you are the sole maintainer.

**Comment by KurtStam** [comment]
@buchdag sorry I misunderstood that. I'm happy to share the maintenance of it :). But seriously this workaround should be gone soon as it's simply an early build of the docker command line utility that includes the manifest capability. The only reason it's wrapped up in a container is so we can run it on travis.

**Comment by buchdag** [comment]
In its current state, your PR actually does not perform any test beside official-images tests.

One obvious reason is that you changed the image tag variable in .travis.yml from `kylemanna/openvpn` to `kylemanna/docker-openvpn`.

The other, less obvious reason is that the image tag has been hardcoded in most of the test instead of using `$1`.

[I ran a version of your PR with everything fixed to have the tests perform as they should](https://travis-ci.org/buchdag/docker-openvpn/builds/319274228), and unfortunately the arm64 image does not pass the test.

**Comment by buchdag** [comment]
I experimented more with this. Turns out you can autobuild and push the alternatives archs without issues straight from Dockerhub if you use [build hooks](https://docs.docker.com/docker-cloud/builds/advanced/). No need to have Travis build and push them at all. The same goes for the manifest.

Details are in the last three commits of this branch:  https://github.com/buchdag/docker-openvpn/commits/aarch64-autobuild

results:
- https://hub.docker.com/r/duch/docker-openvpn-dockerhub/tags/
- https://hub.docker.com/r/duch/docker-openvpn-dockerhub/builds/bdyymf92xht7fxbjt4ssgnl/
- manifest:
```
Name:   duch/docker-openvpn-dockerhub:latest (Type: application/vnd.docker.distribution.manifest.list.v2+json)
Digest: sha256:352f5f1fb6f46624cd226c400c4351a91a3da1d49010dcf834445f51275e2f3e
 * Contains 3 manifest references:
1    Mfst Type: application/vnd.docker.distribution.manifest.v2+json
1       Digest: sha256:7020489e94734443ba2de4619ae07aab495cc72d4d0af0e71f7c8cd8955fad1c
1  Mfst Length: 1363
1     Platform:
1           -      OS: linux
1           - OS Vers: 
1           - OS Feat: []
1           -    Arch: amd64
1           - Variant: 
1           - Feature: 
1     # Layers: 5
         layer 1: digest = sha256:2fdfe1cd78c20d05774f0919be19bc1a3e4729bce219968e4188e7e0f1af679d
         layer 2: digest = sha256:5e151300a1d4a09a67d38e9ce4fac01778f93a7e526e710502d935db4ccdc6ba
         layer 3: digest = sha256:95693d789f2ddccffbac65e2d16fc61c26ef0d70d5d16132a1f6bd19653ff303
         layer 4: digest = sha256:14a2283b08da3155e3f69278f12bf2c201cf81c26ff58937ca949770814bd4ef
         layer 5: digest = sha256:5da3533924baffd2115972f29768254a07fa89e2a1b0db33f8ff67ca82bf9a5e

2    Mfst Type: application/vnd.docker.distribution.manifest.v2+json
2       Digest: sha256:861aa21d71cda7bf9a5e05f78248a52d58ccad8c5ad3434181448aa50d7bfbbc
2  Mfst Length: 1574
2     Platform:
2           -      OS: linux
2           - OS Vers: 
2           - OS Feat: []
2           -    Arch: arm64
2           - Variant: 
2           - Feature: 
2     # Layers: 6
         layer 1: digest = sha256:2b5ec2681cda19b50f9dc500822a39cc5cd1645a230b30301076d90bd371d3fc
         layer 2: digest = sha256:4eded69cf986e38769c49e98662c56394114f2deae96d5c692eda49909e81fed
         layer 3: digest = sha256:8a8b161aeb9a2f3deccf12d374a4f59a61e08bcf2b374a7c54c7bfe9bc17cba9
         layer 4: digest = sha256:a235da92b02025aa3e418f2061408fcdbe86d922ab2f057ac1cfd825a2462660
         layer 5: digest = sha256:5c5c82068960c1846557fb365bfd5bae4998a2e46a62973ab1e905cb2a353add
         layer 6: digest = sha256:10d9cd95cb5ece5d43c0d982d4635c88d4742e4a004f08f7deea6adad448cdc8

3    Mfst Type: application/vnd.docker.distribution.manifest.v2+json
3       Digest: sha256:d08c6b1805d3349e3dc14f8d9e899288b5652c32ebd4ab4d822effff9e6977fc
3  Mfst Length: 1574
3     Platform:
3           -      OS: linux
3           - OS Vers: 
3           - OS Feat: []
3           -    Arch: arm
3           - Variant: 
3           - Feature: 
3     # Layers: 6
         layer 1: digest = sha256:bb18cb19d69048f6a09f02ea479e6006557693e6d60c6b9745b0a05aacf6cd0a
         layer 2: digest = sha256:561ee656703ab2b7e2f77bdf73ba4499e4e872157da76559bc1b4a07ba20ff36
         layer 3: digest = sha256:75e2d5a70bc52e160b194b5c7913a7c72fa97e528f5b42afa4598bb46780a97a
         layer 4: digest = sha256:2b20802225c0b266399ef6c55a6b2c90e23b647495023b1a0feb1fc36310bf61
         layer 5: digest = sha256:d742a5e5ea26d626aeec7d1767b7bac01e77bbcd8d6463996b84b5a8a7683169
         layer 6: digest = sha256:562997ee9ff49c038e56109cdb93d3321342028f2de1cb7b91834ce22a52ad93
```

But unfortunately neither of those two images built from `multiarch/alpine` are working.

The `arm64` image fail test in Travis seemingly due to the kernel of this image being built without iptable nat support.

The `arm` image's `ovpn_genconfig` stalls for unknown reason, so the Travis test time out. I tested this image on a raspberry pi 3 running raspbian (so armv7l), `ovpn_genconfig` never complete like in Travis, and you end up with a non functional configuration.

However It does not appear to be specific to the `multiarch/alpine:arm` image as an image built on the raspberry (so, not using `multiarch/alpine` but vanilla arm alpine from the library) has the exact same issue.

There is maybe something we can do to have at least a working `arm` image. It get stuck (doest not respond to SIGINT) exactly [on this loop](https://github.com/kylemanna/docker-openvpn/blob/fd8daa133aa9ea6b802973b18eccf61272c5a50f/bin/ovpn_genconfig#L302), after echoing `OVPN_CN=vpn.servername.com` to the ovpn_env.sh file.

**Comment by miesgre** [comment]
Nice work @buchdag 
I have working the image on a armhf. You need to patch the `ovpn_genconfig` script. 

you need to change the loop
```
while read -r var; do
  echo "declare -x $var"  >> "$OVPN_ENV"
done < <(set | grep '^OVPN_')
```
to
```
(set | grep '^OVPN_') | while read -r var; do
  echo "declare -x $var"  >> "$OVPN_ENV"
done
```
Patch here (https://github.com/miesgre/docker-openvpn/commit/23aeadb9ab46338337fb86a773ff20676782561a)

The problem is that it fails to make the tests in travis  (https://travis-ci.org/miesgre/docker-openvpn/jobs/319586482) because  the `paranoid` test takes to long without writing to the stdout. It is generating the certificates and in arm it takes a lot of time. 

I tried to use the `travis_wait` as documented in [the travis doc](https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received) with no results (https://travis-ci.org/miesgre/docker-openvpn/jobs/319624433), but you can check that the `paranoid` test passed!

Maybe the tests can be modified to write to the stdout while generating the certificates? I think this will solve the problem.

**Comment by buchdag** [comment]
I think most of the tests will have to be modified anyway if we want to tests for non `amd64` archs on Travis using qemu.

There are a lot of redundant and very expensive crypto operation in them (mostly 2048 bits DH parameters generation) and all of them have the potential to trigger either the 10 minutes without write to std out time out individually or the 50 minutes global time out when put together.

**Comment by miesgre** [comment]
I split the test and some tests does not have enough time to finish and others does not pass
(https://travis-ci.org/miesgre/docker-openvpn/builds/320219462)

**Comment by buchdag** [comment]
Looks like `multiarch/alpine:armhf-v3.7` has (at least) the same iptables issue: https://travis-ci.org/miesgre/docker-openvpn/jobs/320219477#L645 .

Do you know if it's the same in library's alpine 3.7 ?

**Comment by miesgre** [comment]
@buchdag I think so

All images `alpine:latest`, `multiarch/alpine:armhf-v3.7` and `multiarch/alpine:arm64-v3.7` installs the same library version
```
(16/23) Installing iptables (1.6.1-r1)
```

**Comment by buchdag** [comment]
Unfortunately this is not related to the iptables library used but to a missing kernel compilation option. By library I meant Dockerhub's Library, ie the image available through Dockerhub without a repo prefix (nginx, alpine, debian, etc).

There is  no `/proc/config.gz` file in library's arm version of `alpine:3.7` nor in `multiarch/alpine:armhf-v3.7`, meaning that the kernel was compiled without `CONFIG_IKCONFIG_PROC` (it's not the case for the `amd64` version of `alpine:3.7`) so we can't even check what options were used.

I'm afraid the only people who can fix the missing kernel compile option for iptables nat support are multiarch or alpine maintainers.

**Comment by miesgre** [comment]
I think the problem is in the qemu used in travis to run the tests. Running on arm hardware seams to work.

Running `$ iptables -L -t nat` in a Raspberry Pi 3 (it works!):
```
$ docker run -it --cap-add=NET_ADMIN multiarch/alpine:armhf-v3.7 sh -c "apk add iptables --update; iptables -L -t nat"
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/armhf/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/armhf/APKINDEX.tar.gz
(1/4) Installing libmnl (1.0.4-r0)
(2/4) Installing jansson (2.10-r0)
(3/4) Installing libnftnl-libs (1.0.8-r1)
(4/4) Installing iptables (1.6.1-r1)
Executing busybox-1.27.2-r7.trigger
OK: 6 MiB in 20 packages
Chain PREROUTING (policy ACCEPT)
target     prot opt source               destination

Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination

Chain POSTROUTING (policy ACCEPT)
target     prot opt source               destination
```

Running the same in a X86 with the qemu (like in travis-ci, does not work!):
```
$ docker run -it --cap-add=NET_ADMIN multiarch/alpine:armhf-v3.7 sh -c "apk add iptables --update; iptables -L -t nat"
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/main/armhf/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.7/community/armhf/APKINDEX.tar.gz
(1/4) Installing libmnl (1.0.4-r0)
(2/4) Installing jansson (2.10-r0)
(3/4) Installing libnftnl-libs (1.0.8-r1)
(4/4) Installing iptables (1.6.1-r1)
Executing busybox-1.27.2-r7.trigger
OK: 6 MiB in 20 packages
iptables v1.6.1: can't initialize iptables table `nat': iptables who? (do you need to insmod?)
Perhaps iptables or your kernel needs to be upgraded.
```

$ zcat /proc/config.gz | grep IPTABLES       
```
$ docker run -it --cap-add=NET_ADMIN multiarch/alpine:armhf-v3.7 zcat /proc/config.gz | grep IPTABLES       
CONFIG_IP_NF_IPTABLES=y
CONFIG_IP6_NF_IPTABLES=y
```

$ zcat /proc/config.gz | grep FILTER
```
$ docker run -it --cap-add=NET_ADMIN multiarch/alpine:armhf-v3.7 zcat /proc/config.gz | grep FILTER
CONFIG_HAVE_ARCH_SECCOMP_FILTER=y
CONFIG_SECCOMP_FILTER=y
CONFIG_NETFILTER=y
# CONFIG_NETFILTER_DEBUG is not set
CONFIG_NETFILTER_ADVANCED=y
CONFIG_BRIDGE_NETFILTER=y
CONFIG_NETFILTER_INGRESS=y
CONFIG_NETFILTER_NETLINK=y
CONFIG_NETFILTER_NETLINK_ACCT=y
CONFIG_NETFILTER_NETLINK_QUEUE=y
CONFIG_NETFILTER_NETLINK_LOG=y
CONFIG_NETFILTER_NETLINK_GLUE_CT=y
CONFIG_NETFILTER_SYNPROXY=y
CONFIG_NETFILTER_XTABLES=y
CONFIG_NETFILTER_XT_MARK=y
CONFIG_NETFILTER_XT_CONNMARK=y
CONFIG_NETFILTER_XT_SET=y
....
```

**Comment by buchdag** [comment]
Agreed, I came to the same conclusion yesterday.

[I lowered the rsa keysize in tests for `arm` archs to remove the timeout issue](https://travis-ci.org/buchdag/docker-openvpn/builds/322407562), test still failing are `basic`, `dual-proto`, `otp`, `iptables` and `revocation`, in short all tests that have iptables running inside the container.

I tried to bypass the `iptables` test and pass the `-d` arg (Disable default route) to `ovpn_config` to avoid running iptables at all on `arm` archs, [the tests now fails due to an ioctl issue related to the TUN/TAP interface](https://travis-ci.org/buchdag/docker-openvpn/jobs/322401830) (qemu again I guess).

I think we are reaching the limit of what is currently possible (or possible at all) with Travis and Qemu.

On a sidenote, due to the fact that alpine like most Library images is now multiarch using a manifest, the existing Dockerfile builds fine on `arm` (and probably on `arm64` as well) without any modification. I just started the whole test suite on a Pi 3 to see if all tests pass when running on real `arm` arch.

**Comment by buchdag** [comment]
All test passed on an image built straight from the current Dockerfile on a Raspberry Pi 3.

It took four hours and a half.

**Comment by miesgre** [comment]
@buchdag I changed the base images from `multiarch/alpine` to the official ones and build is working correctly on docker hub (https://hub.docker.com/r/miesgre/openvpn/tags/).

Tests on travis are still failing. I think the best is to disable all iptables tests in non amd64 arch.

I'm sending you a PR with the changes I done to use the official alpine images instead multiarch ones.

**Comment by buchdag** [comment]
> I think the best is to disable all iptables tests in non amd64 arch.

Well, they're not really iptables tests, only one of them is explicitly testing iptables, the others test the very core functionality of the container, disabling them would be pretty much the same as not testing the images at all.

Plus as mentioned earlier I tried to get rid of the call to iptables and still hit yet another error (an ioctl one).

**Comment by buchdag** [comment]
> All test passed on an image built straight from the current Dockerfile on a Raspberry Pi 3.

Ditto on a low end arm64 host. Took 20 minutes.

What we could do for now is remove `Dockerfile.aarch64` as it is no longer needed and say in the doc that the existing Dockerfile can be used to build the image on other archs (and appears to work at least on `arm` and `arm64`), but that due to the current inability to fully test those other archs with Travis CI, they won't be pushed to dockerhub.

**Comment by buchdag** [comment]
@kylemanna any advice on this (the last paragraph) ?

**Comment by KurtStam** [comment]
@buchdag @kylemanna  How about running a subset of the tests and mark the docker hub images as alpha? Then opening new issues to get the remaining tests running? Seems to me it'd be nice to move forward.

**Pull request closed without merging**